### PR TITLE
Tutorial about NaNs

### PR DESCRIPTION
**Pull request by hantek**
This is a tentative tutorial about how to deal with NaNs. If you have more suggestions, please let me know so that I can add them into the tutorial. 

(A reply for #2770)

### CODE REVIEW

**Comment by nouiz** [comment]
nanguard isn't yet in Theano. So before merging this PR that talk about it,
can you move it in Theano in this PR?

**Comment by hantek** [comment]
Yes, sure!

**Comment by nouiz** [comment]
I think they where tests for that mode, also include them.

**Comment by hantek** [comment]
Hi, I guess it is well now.

File: doc/library/compile/nanguardmode.txt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+   :synopsis: defines NanGuardMode
+.. moduleauthor:: LISA
+
+Guide
+=====
+
+
+The NanGuardMode aims to prevent the model from outputing NaNs or Infs. It has
+a number of self-checks, which can help to find out which apply node is
+generating those incorrect output.
```
Comment (dwf) [inline doc/library/compile/nanguardmode.txt]: outputs*

File: doc/tutorial/nan_tutorial.txt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+.. _nan_tutorial:
+
+=================
+Dealing with NaNs
+=================
+
+Having a model yielding NaNs or Infs is quite common if some of the tiny
+components in your model are not set properly. NaNs are hard to deal with
+because sometimes it is caused by a bug or error in the code, sometimes it's
+because of the numerical stability of your actual computing systerm, and even,
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: computing systerm -> algorithm

Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: Actually, I don't know what you mean by "computing system", I see you mention algorithm below.

File: doc/tutorial/nan_tutorial.txt [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+sometimes it relates to your algorithm. Here we try to outline common, basic
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: remove ", basic"

File: doc/tutorial/nan_tutorial.txt [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+issues which cause the model to yield NaNs, as well as provide nails and
+hammers to diagnose it. 
+
+
+Check Superparameters and Weight Initialization
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: Hyperparameters?

File: doc/tutorial/nan_tutorial.txt [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+-----------------------------------------------
+
+Most frequently, the cause would be that some of the superparameters, especially
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: hyperparameters

File: doc/tutorial/nan_tutorial.txt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+learning rates, are set incorrectly. A high learning rate can blow up your whole
+model into NaN outputs even within one epoch of training. So the first and
+easiest way is try to lower it. Keep halving your learning rate until you start
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: way -> solution

File: doc/tutorial/nan_tutorial.txt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+to get resonable output values.
+
+Other superparameters may also play a role. For example, are your training
+algorithms involve regularization terms? If so, are their corresponding
+penalties set resonably? Search a wider superparameter space with a few (one or
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: resonably -> reasonably

File: doc/tutorial/nan_tutorial.txt (hunk 5 above)
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: In fact just search and replace superparameter with hyperparameter, that's the more common name

File: doc/tutorial/nan_tutorial.txt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+that the model ends up with yielding NaNs. 
+
+
+Run in DebugMode
+-----------------
+
+If adjusting superparameters doesn't work for you, you can still get help from
+Theano's DebugMode. Run your code in DebugMode with flag
+DebugMode.check_py=False. This will give you clue about which op is causing this
+problem, and then you can inspect into that op in more detail.
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: Do you also not need mode=DebugMode? Maybe give a full example of the THEANO_FLAGS to use

File: doc/tutorial/nan_tutorial.txt [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+executed when the function is called. For how to use that, please check
+:ref:`faq_monitormode`. 
+
+
+Numerical Stability
+-------------------
+
+After you have located the op which causes the problem, it may turn out that the
+NaNs yielded by that op are related to numerical issues. For example,
+1 / log(p(x) + 1) may result in NaNs for those nodes who have learned to yield
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: put `:math:\`1 / log(p(x) + 1)\``

File: doc/tutorial/nan_tutorial.txt [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+a low probability p(x) for some input x.
+
+
+Algorithm Related
+-----------------
+
+The hardest thing is that, after tracing back through all the former processes,
```
Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: I'd rephrase this:

"In the most difficult situations, you may go through the above steps and find nothing wrong. If the above methods fail to uncover the cause, there is a good chance that something is wrong with your algorithm." Next sentence is fine.

**Comment by dwf** [comment]
Sorry I took so long for feedback.

File: doc/tutorial/nan_tutorial.txt (hunk 2 above)
Comment (hantek) [inline doc/tutorial/nan_tutorial.txt]: It is the environmental configuration, like numpy, scipy, etc. I would put "your numpy version" instead.

Comment (dwf) [inline doc/tutorial/nan_tutorial.txt]: I would say "your computational environment (library versions, etc.)"
